Compilation errors in `Schema.d.ts` file

## 🐛 Bug report

### Current Behavior

<!-- If applicable, add screenshots to help explain your problem. -->
Running `tsc` for any code using `@effect/schema` library version `0.9.0` results in the following error messages:
```
node_modules/@effect/schema/Schema.d.ts:273:324 - error TS2536: Type 'K_4' cannot be used to index type '{ readonly [K in Exclude<keyof Fields, OptionalKeys<Fields>>]: From<Fields[K]>; } & { readonly [K_1 in OptionalKeys<Fields>]?: From<Fields[K_1]>; }'.

273 export declare const struct: <Fields extends Record<PropertyKey, Schema<any, any> | OptionalSchema<any, any>>>(fields: Fields) => Schema<{ readonly [K in Exclude<keyof Fields, OptionalKeys<Fields>>]: From<Fields[K]>; } & { readonly [K_1 in OptionalKeys<Fields>]?: From<Fields[K_1]>; } extends infer T ? { [K_4 in keyof T]: ({ readonly [K in Exclude<keyof Fields, OptionalKeys<Fields>>]: From<Fields[K]>; } & { readonly [K_1 in OptionalKeys<Fields>]?: From<Fields[K_1]>; })[K_4]; } : never, { readonly [K_2 in Exclude<keyof Fields, OptionalKeys<Fields>>]: To<Fields[K_2]>; } & { readonly [K_3 in OptionalKeys<Fields>]?: To<Fields[K_3]>; } extends infer T_1 ? { [K_5 in keyof T_1]: ({ readonly [K_2 in Exclude<keyof Fields, OptionalKeys<Fields>>]: To<Fields[K_2]>; } & { readonly [K_3 in OptionalKeys<Fields>]?: To<Fields[K_3]>; })[K_5]; } : never>;

                                                                                                                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@effect/schema/Schema.d.ts:273:681 - error TS2536: Type 'K_5' cannot be used to index type '{ readonly [K_2 in Exclude<keyof Fields, OptionalKeys<Fields>>]: To<Fields[K_2]>; } & { readonly [K_3 in OptionalKeys<Fields>]?: To<Fields[K_3]>; }'.

273 export declare const struct: <Fields extends Record<PropertyKey, Schema<any, any> | OptionalSchema<any, any>>>(fields: Fields) => Schema<{ readonly [K in Exclude<keyof Fields, OptionalKeys<Fields>>]: From<Fields[K]>; } & { readonly [K_1 in OptionalKeys<Fields>]?: From<Fields[K_1]>; } extends infer T ? { [K_4 in keyof T]: ({ readonly [K in Exclude<keyof Fields, OptionalKeys<Fields>>]: From<Fields[K]>; } & { readonly [K_1 in OptionalKeys<Fields>]?: From<Fields[K_1]>; })[K_4]; } : never, { readonly [K_2 in Exclude<keyof Fields, OptionalKeys<Fields>>]: To<Fields[K_2]>; } & { readonly [K_3 in OptionalKeys<Fields>]?: To<Fields[K_3]>; } extends infer T_1 ? { [K_5 in keyof T_1]: ({ readonly [K_2 in Exclude<keyof Fields, OptionalKeys<Fields>>]: To<Fields[K_2]>; } & { readonly [K_3 in OptionalKeys<Fields>]?: To<Fields[K_3]>; })[K_5]; } : never>;



                                                                                                                                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

node_modules/@effect/schema/Schema.d.ts:433:245 - error TS2536: Type 'K_1' cannot be used to index type 'A & { readonly [k in K]: V; }'.

433 export declare const attachPropertySignature: <K extends PropertyKey, V extends AST.LiteralValue>(key: K, value: V) => <I, A extends object>(schema: Schema<I, A>) => Schema<I, A & { readonly [k in K]: V; } extends infer T ? { [K_1 in keyof T]: (A & { readonly [k in K]: V; })[K_1]; } : never>;

                                                                            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 3 errors in the same file, starting at: node_modules/@effect/schema/Schema.d.ts:273
```

### Expected behavior

<!-- A clear and concise description of what you expected to happen. -->
Running `tsc` will not produce any errors.

### Reproducible example
Any code which `import`s `@effect/schema` library version `0.9.0`.


### Suggested solution(s)

<!-- How could we solve this bug? What changes would need to made to `@effect/schema`? -->
I [again propose](https://github.com/Effect-TS/io/issues/324#issuecomment-1492495989) adding a step to your CI/CD pipeline, which would try to compile/transpile some small dummy code against the `d.ts` files produced by previous compilation/transpilation step.
That will add safeguard against these kind of errors for the future, at least for one TSC version.

### Additional context

<!-- Add any other context about the problem here.  -->
There luckily is workaround to add `--skipLibCheck` to `tsc`, so this bug is not a blocker.
Though, the workaround will then affect **all** `.d.ts` files, both own and all the libraries in `node_modules`.

### Your environment

<!-- PLEASE FILL THIS OUT -->

| Software      | Version(s) |
| ------------- | ---------- |
| @effect/schema |  0.9.0          |
| TypeScript    |     5.0.2       |

I think I saw similar error also in `0.8.0`.

--- a/.changeset/hot-wombats-attack.md
+++ b/.changeset/hot-wombats-attack.md
@@ -1,0 +1,5 @@
+---
+"@effect/schema": patch
+---
+
+use `Simplify` from `effect/Types` in `TaggedClass`, `TaggedError` and `TaggedRequest` to avoid errors in `Schema.d.ts`, closes #1841

--- a/packages/schema/src/Schema.ts
+++ b/packages/schema/src/Schema.ts
@@ -32,7 +32,7 @@ import * as Request from "effect/Request"
 import * as _secret from "effect/Secret"
 import * as SortedSet from "effect/SortedSet"
 import * as S from "effect/String"
-import type { Covariant, Invariant, NoInfer, UnionToIntersection } from "effect/Types"
+import type * as Types from "effect/Types"
 import type { Arbitrary } from "./Arbitrary.js"
 import * as arbitrary from "./Arbitrary.js"
 import type { ParseOptions } from "./AST.js"
@@ -143,9 +143,9 @@ export declare namespace Schema {
    */
   export interface Variance<A, I, R> {
     readonly [TypeId]: {
-      readonly _A: Invariant<A>
-      readonly _I: Invariant<I>
-      readonly _R: Covariant<R>
+      readonly _A: Types.Invariant<A>
+      readonly _I: Types.Invariant<I>
+      readonly _R: Types.Covariant<R>
     }
   }
 
@@ -1833,7 +1833,7 @@ export declare namespace IndexSignature {
    */
   export type Type<
     Records extends IndexSignature.Records
-  > = UnionToIntersection<
+  > = Types.UnionToIntersection<
     {
       [K in keyof Records]: {
         readonly [P in Schema.Type<Records[K]["key"]>]: Schema.Type<Records[K]["value"]>
@@ -1846,7 +1846,7 @@ export declare namespace IndexSignature {
    */
   export type Encoded<
     Records extends IndexSignature.Records
-  > = UnionToIntersection<
+  > = Types.UnionToIntersection<
     {
       [K in keyof Records]: {
         readonly [P in Schema.Encoded<Records[K]["key"]>]: Schema.Encoded<Records[K]["value"]>
@@ -2449,8 +2449,8 @@ export function filter<C extends A, B extends A, A = C>(
   annotations?: Annotations.Filter<A>
 ): <I, R>(self: Schema<C, I, R>) => Schema<C & B, I, R>
 export function filter<A>(
-  predicate: Predicate.Predicate<NoInfer<A>>,
-  annotations?: Annotations.Filter<NoInfer<A>>
+  predicate: Predicate.Predicate<Types.NoInfer<A>>,
+  annotations?: Annotations.Filter<Types.NoInfer<A>>
 ): <I, R>(self: Schema<A, I, R>) => Schema<A, I, R>
 export function filter<A>(
   predicate: Predicate.Predicate<A> | AST.Refinement["filter"],
@@ -6042,10 +6042,10 @@ export const TaggedClass = <Self = never>(identifier?: string) =>
   : Class<
     Self,
     { readonly [TAG]: literal<[Tag]> } & Fields,
-    Simplify<{ readonly [TAG]: Tag } & Struct.Type<Fields>>,
-    Simplify<{ readonly [TAG]: Tag } & Struct.Encoded<Fields>>,
+    Types.Simplify<{ readonly [TAG]: Tag } & Struct.Type<Fields>>,
+    Types.Simplify<{ readonly [TAG]: Tag } & Struct.Encoded<Fields>>,
     Struct.Context<Fields>,
-    Simplify<Struct.Type<Fields>>,
+    Types.Simplify<Struct.Type<Fields>>,
     {},
     {}
   > =>
@@ -6073,10 +6073,10 @@ export const TaggedError = <Self = never>(identifier?: string) =>
   : Class<
     Self,
     { readonly [TAG]: literal<[Tag]> } & Fields,
-    Simplify<{ readonly [TAG]: Tag } & Struct.Type<Fields>>,
-    Simplify<{ readonly [TAG]: Tag } & Struct.Encoded<Fields>>,
+    Types.Simplify<{ readonly [TAG]: Tag } & Struct.Type<Fields>>,
+    Types.Simplify<{ readonly [TAG]: Tag } & Struct.Encoded<Fields>>,
     Struct.Context<Fields>,
-    Simplify<Struct.Type<Fields>>,
+    Types.Simplify<Struct.Type<Fields>>,
     {},
     Cause.YieldableError
   > =>
@@ -6131,10 +6131,10 @@ export const TaggedRequest =
     : Class<
       Self,
       { readonly [TAG]: literal<[Tag]> } & Fields,
-      Simplify<{ readonly [TAG]: Tag } & Struct.Type<Fields>>,
-      Simplify<{ readonly [TAG]: Tag } & Struct.Encoded<Fields>>,
+      Types.Simplify<{ readonly [TAG]: Tag } & Struct.Type<Fields>>,
+      Types.Simplify<{ readonly [TAG]: Tag } & Struct.Encoded<Fields>>,
       Struct.Context<Fields>,
-      Simplify<Struct.Type<Fields>>,
+      Types.Simplify<Struct.Type<Fields>>,
       TaggedRequest<
         Tag,
         Self,

